### ISSUE TITLE
Site doesn't render ok when serving due to browser cache issues, what's the recommended way to go?

### ISSUE THREAD

**Issue by leworking** [gatsbyjs/gatsby#12376]
EDITED

--------------------------------
_The problem_, after a `build`, the site doesn't render well when `served`, yes when `developed` and when `deployed on Netlify`

So it works well when `serve` if going with an "incognito" mode. 
It is strange because on my non-incognito Chrome, reading a value from `context` (and posting in `console.log`) it gives a value that is no longer in the code (so it hasn't been build). Who is caching these values, the `service worker`?

If so, how is it solved? Emptying the cache of the browser? Will this happen to regular users when I update the site and send more deployments?

--------------------------------

I have finished a site that runs as expected in `develop`, that I can `build` and deploy to Netlify by dropping the `public` folder, and it works as expected, but that if I `serve` (from the same build) I see things like:

- sometimes the web page appears, but with no images
- sometimes the page is blank, without any real html content as seen in `inspect` (see the screenshot below)
- sometimes the page gets blank not from the beginning but when navigating within the site

The `build` process doesn't complain, I have deleted before the `cache` and `public` folders just in case. I've repeated the process a couple of times and the results are consistent

And I do see error messages in the console when `serving` it

`Error: An error occurred. See https://github.com/styled-components/styled-components/blob/master/src/utils/errors.md#8 for more information. Additional arguments: `

that **I don't see on the deployed site**

I'm gonna troubleshoot the `styled-components` message, but obviously the fact that everything goes all right on the `develop` and `build & deployed` site makes me suspicious. **Any suggestion?**

![image](https://user-images.githubusercontent.com/4474004/53958148-8d5c5000-40e0-11e9-9e6e-bd86d3117b91.png)

**Comment by leworking** [comment]
I haven't been able to make it run by bypassing the `service workers`, deleting, updating them ...
But I have been able by going to `localStorage`, which I access in the `site` through `context`, now through _Chrome Inspect_, and there I've seen the old value that I don't know why is still there. 

I've manually changed it to the value it should have and the site now renders ok.
If I change it again to another random value, the site renders bad again.

If I delete the entry, then the site goes alive again, and I can see I can change the value through the web site without any problem.

If I change it again, then again it gets blocked, the site doesn't render well, and it seems not to be able to change that value (but probably it is that javascript is not running well)

So my **conclusion** is that somehow the site didn't have access to change that value(??) how can that be? I'm a little bit confused right now O_O

**Comment by leworking** [comment]
At the end, the problem was that if the variable existed in localStorage with a value that wasn't expected, this created an error that bubbled up to an absence of rendering.

This happened only in build because by chance, there was still an old localStorage value there. If I manually changed the value in the other environments I reproduced the error without problems.

I've solved it by checking that the value is one of the expected ones, and if not changing it, so at the end it wasn't a problem of gatsby (apart from the fact that error messages are helpless indeed). So I close the ticket, and edit the title to better reflect what was this about.
